### ISSUE TITLE
Memory leak when using DX12 backend in debug configuration

### ISSUE THREAD

**Issue by savant117** [DiligentGraphics/DiligentCore#101]
When running the tutorials in debug configuration with the dx12 backend, there is a pretty severe memory leak. It can be seen with any of the tutorial projects, but none of the samples. I haven't dug too deeply so I'm not sure if its a problem with the samples or the backend itself.

Steps to reproduce:
1) Check out fresh copy of the diligent engine repository
2) Build and run any of the tutorial projects in debug configuration
3) Select DX12 as the backend
4) Observe the memory usage climb until crash

My setup for reference:
Windows 10 64 bit, RTX 2070, visual studio 2019 community

**Comment by DiligentGraphics** [comment]
Is this only happening in debug mode? Did you see the leaks with other backends? How fast does the memory leak?
Debug mode runs with crt memory leak reporting on, so there would be message if it was something obvious.
It is strange this only happens with tutorials as they use the same infrastructure as samples.

**Comment by DiligentGraphics** [comment]
Does the memory also leak in Tutorial00 that doesn't use any infrastructure and is a very basic Win32 app?

**Comment by savant117** [comment]
I tested all backends in both release and debug mode for the tutorials. It happens for only with dx12 backend, and only in debug mode (not release). It does happen in Tutorial00. It leaks at about 20 - 50 MB/s

**Comment by DiligentGraphics** [comment]
I would've definitely noticed that issue if that was happening on my machine. I had been rinning Tutorial15 for a good hour recently to test simulation and it worked just fine.
Also the fact that it happens in debug mode only seems to be suspicious.

Can you try few thing in Tutorial00: 
- disable rendering
- disable rendering and presenting
- do not create debug device

**Comment by savant117** [comment]
If I comment out Render() on line 488 I get the following:
![image](https://user-images.githubusercontent.com/7024002/63190419-2c905680-c034-11e9-8fec-937f3a7957d7.png)

If I comment out Render and Present lines 488 - 489, the window is blank but no more memory leak.

How do I not create the debug device?

**Comment by DiligentGraphics** [comment]
Try to comment everything debug-related here:
https://github.com/DiligentGraphics/DiligentCore/blob/master/Graphics/GraphicsEngineD3D12/src/EngineFactoryD3D12.cpp

You will need to rebuild the Tutorial for it to pick up updated dlls

**Comment by savant117** [comment]
Confirmed - commenting out debug device stuff fixes the leak

**Comment by DiligentGraphics** [comment]
I've seen memory leaks in Vulkan validation layers, but at least on my version of Windows, d3d12 debug stuff works fine. I'm wondering if you can reproduce it with some simple d3d12 sample from MSFT.

**Comment by savant117** [comment]
Ok, so I tried running the official samples from here:
https://github.com/microsoft/DirectX-Graphics-Samples/tree/master/Samples/Desktop/D3D12HelloWorld

But I am not able to reproduce it. I'm not very familiar with the DX12 API but I will try to dig a little deeper. Let me know if you have any other ideas for me to try.

**Comment by DiligentGraphics** [comment]
Can you try to copy the code you commented to the sample where it creates the device? It enables some message filtering which could be the cause

**Comment by savant117** [comment]
Tried pasting that in the sample, and tried commenting just that bit out in diligent. No leak in d3d sample, still leaking in diligent

**Comment by savant117** [comment]
Not sure if this helps but, if I comment out the ClearDepthStencil() in the render function, the memory usage goes up at half the rate. So maybe the leak is in ClearDepthStencil() and / or ClearRenderTarget()

**Comment by DiligentGraphics** [comment]
Hm... Samples and tutorials use the same infrastructure, so there is no reasonable difference I can think of. Will take a look

**Comment by DiligentGraphics** [comment]
There's nothing in ClearDepthStencil that would behave differently in debug vs release modes.

BTW, Diligent reports resource usage when you exits the app. Can you post it here?

**Comment by savant117** [comment]
OK. So, when I run any of the tutorials, clicking the windows (x) closes the window and exits the application cleanly *Except* when using dx12 in debug. If I click the X, the window will close but the application does not exit. Visual studio debugger is still active when the window is closed. However, the memory usage stops going up.

Heres the print out for running the application for a few seconds then closing the window:

Diligent Engine: Info: User-defined allocator is not provided. Using default allocator.
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x19845000
'Tutorial01_HelloTriangle.exe' (Win32): Loaded 'C:\Windows\System32\dcomp.dll'. 
'Tutorial01_HelloTriangle.exe' (Win32): Loaded 'C:\Windows\System32\dxilconv.dll'. 
D3D12 ERROR: ID3D12Device::CopyDescriptorsSimple: Source ranges and dest ranges overlap, which results in undefined behavior. [ EXECUTION ERROR #653: COPY_DESCRIPTORS_INVALID_RANGES]
D3D12 ERROR: ID3D12Device::CopyDescriptorsSimple: Source ranges and dest ranges overlap, which results in undefined behavior. [ EXECUTION ERROR #653: COPY_DESCRIPTORS_INVALID_RANGES]
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x19a26000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1ad00000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1ae00000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1af20000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1b0e0000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1b200000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1b3a0000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1b500000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1b680000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1b7c0000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1b8c0000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1bb00000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1bc40000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1bd60000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1bf00000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1c140000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1c280000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1c3a0000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1c4e0000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1c7c0000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1c8e0000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1ca20000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1cc00000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1ce00000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1cf20000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1d060000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1d180000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1d2c0000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1d560000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1d6a0000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1d7c0000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1d900000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1da20000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1dce0000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1de00000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1df40000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1e060000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1e1a0000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1e2c0000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1e400000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1e560000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1e840000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1ea00000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1eb80000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1ecc0000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1ede0000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1ef20000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1f040000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1f180000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1f2a0000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1f3e0000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1f500000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1f960000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1faa0000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1fbc0000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1fd00000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x1fe20000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x280200000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x280380000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x280600000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x280820000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x280a40000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x280b60000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x280ca0000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x280dc0000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x280f00000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x281020000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x281160000
Diligent Engine: Info: Created dynamic memory page. Size: 1.00 MB; GPU virtual address 0x281280000
Diligent Engine: Info: SAMPLER     dynamic descriptor allocator of immediate context usage stats: peak descriptor count: 0/0
Diligent Engine: Info: CBV_SRV_UAV dynamic descriptor allocator of immediate context usage stats: peak descriptor count: 0/0
Diligent Engine: Info: Dynamic heap of immediate context usage stats:
                       Peak used/aligned/allocated size: 5.97 KB / 7.75 KB / 1.00 MB (1 page). Peak efficiency (used/aligned): 77.0%. Peak utilization (used/allocated): 0.6%

**Comment by DiligentGraphics** [comment]
Hmm.. There is some D3D12 error that promises undefined behavior after which it starts to allocate a lot of memory. Can you share your Windows version and also Windows SDK version you use?

**Comment by savant117** [comment]
Windows 1809 (build 17763.678)
SDK version 10.0.17763.0

**Comment by jayrulez** [comment]
I'm also observing this memory leak. There's also a lot of flickering. I've only tested the Hello Triangle tutorial.

**Comment by DiligentGraphics** [comment]
I can't reproduce the problem. I updated Windows to 10.0.18362 and rebuilt the project with 10.0.17763.0 Windows SDK and everything works fine.
There are few differences though:
- I use Visual Studio 2017. For some reason CMake (3.14.5) fails to generate solution for VS2019 
- My system has GTX970 GPU, so there may be some issue there to

@savant117 Do you see this error when you run other tutorials as well?

```
D3D12 ERROR: ID3D12Device::CopyDescriptorsSimple: Source ranges and dest ranges overlap, which results in undefined behavior. [ EXECUTION ERROR #653: COPY_DESCRIPTORS_INVALID_RANGES]
```

And just for sanity: it does not happen when you run samples?

**Comment by DiligentGraphics** [comment]
Actually, it in fact reproduces in Tutorial00 now, but not in other tutorials

**Comment by jayrulez** [comment]
I do not observe the leak in the samples. However, I still observe the flickering on DX12 so that may be an unrelated issue.

For the samples, the flickering only occurs in the Atmosphere one.

Tutorials
HelloWin32 - No flickering but memory leaks very fast. Uses 1+GB in less than 30 seconds.
HelloTriangle - Flickering and memory leak
Cube - Flickering and memory leak
Texturing - Flickering and memory leak
Instancing - Flickering and memory leak (but leaks much slower)
Texture Array - Flickering and memory leak (but leaks much slower)
Multithreading - Flickering and memory leak. 

I didn't check further as I suspect I'll just be observing more of the same.

I do not observe that specific error in the output but I saw this in the last test I ran
```
Exception thrown at 0x00007FFD8FEDA839 in Tutorial06_Multithreading.exe: Microsoft C++ exception: _com_error at memory location 0x000000A01BDE4640.
```

This is all only happening in for debug builds.

I'll do checks on the release builds next.

**Comment by DiligentGraphics** [comment]
What I found so far makes absolutely no sense: the issue reproduces when
- Debug layer is enabled
- Present interval is 1

When setting present interval to 0 with debug layer enabled, it works fine. When disabling debug layer when presenting with interval 1, it also works fine.

**Comment by DiligentGraphics** [comment]
Because only Tutorial00 uses present interval 1 by default, the issue does not reproduce by default in other tutorials unless I manually set present interval to 1

**Comment by DiligentGraphics** [comment]
So I managed to reproduce the issue in MSFT's D3D12HelloTriangle sample and it does seem to be a bug. Will file an issue.

**Comment by savant117** [comment]
Yes, I do see that with the other tutorials, and I do see it with the samples also

**Comment by DiligentGraphics** [comment]
If you want to reproduce the problem in D3D12HelloTriangle sample from MSFT, modify `OnRender()` function as follows:

```cpp
void D3D12HelloTriangle::OnRender()
{
    //// Record all the commands we need to render the scene into the command list.
    //PopulateCommandList();

    //// Execute the command list.
    //ID3D12CommandList* ppCommandLists[] = { m_commandList.Get() };
    //m_commandQueue->ExecuteCommandLists(_countof(ppCommandLists), ppCommandLists);

    //// Present the frame.
    ThrowIfFailed(m_swapChain->Present(1, 0));

    //WaitForPreviousFrame();
}
```

**Comment by DiligentGraphics** [comment]
I reported the problem to MSFT. Usually, they are reasonably fast to respond.

**Comment by savant117** [comment]
Yep, that does indeed reproduce the problem in D3D12HelloTriangle for me

**Comment by DiligentGraphics** [comment]
Microsoft says this issues should be fixed in insider fast build 18963.

**Comment by TheMostDiligent** [comment]
This bug is fixed in Windows Version 10.0.19041
